tctl-snap/tctl-plugins/cmd/tctl-login/login.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2022 Canonical Ltd.
+
+package main
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/canonical/charmed-temporal-image/tctl-plugins/cmd"
+	"github.com/google/uuid"
+	"github.com/skratchdot/open-golang/open"
+)
+
+const (
+	redirectURI  = "http://localhost:7070/oauth2callback"
+	authEndpoint = "https://accounts.google.com/o/oauth2/auth"
+	scope        = "openid profile email"
+)
+
+var (
+	server       *http.Server
+	shutdown     = make(chan struct{}) // Custom channel for graceful shutdown signal
+	state        = uuid.New().String()
+	clientID     string
+	clientSecret string
+	codeVerifier string
+)
+
+func main() {
+	env := os.Getenv("TCTL_ENVIRONMENT")
+	if env == "dev" {
+		fmt.Fprintf(os.Stderr, "login command not defined for development environment.\n")
+		return
+	}
+
+	var err error
+	clientID, err = cmd.ClientID()
+	if err != nil {
+		return
+	}
+
+	if clientID == "" {
+		fmt.Fprintf(os.Stderr, "no google-client-id found for %v environment. use 'sudo snap set tctl %v-google-client-id=\"<client_id>\"'.\n", env, env)
+		return
+	}
+
+	clientSecret, err = cmd.ClientSecret()
+	if err != nil {
+		return
+	}
+
+	if clientSecret == "" {
+		fmt.Fprintf(os.Stderr, "no google-client-secret found for %v environment. use 'sudo snap set tctl %v-google-client-secret=\"<client_secret>\"'.\n", env, env)
+		return
+	}
+
+	token, _ := cmd.FetchValidToken(clientID, clientSecret)
+	if token != "" {
+		os.Exit(1)
```

Comment:
log maybe (and below too)
Sometimes you return sometimes exit(1), maybe choose one pattern